docs/testing_servers.md

```diff
@@ -0,0 +1,340 @@
+# Testing Servers
+
+Testing servers with the conformance runner involves the following steps:
+
+* Implementing the ConformanceService endpoints.
+* Making an executable file that will read messages from `stdin` and write messages to `stdout`.
+* Defining any configuration for what your server supports. For more information on how to do this, see the [Configuration Files](./configuring_and_running_tests.md#configuration-files) section in the [Configuring and Running Tests documentation](./configuring_and_running_tests.md).
+
+## Process
+
+The basic process for working with servers in the conformance suite will work as follows:
+
+* The conformance runner will pass a `ServerCompatRequest` message via `stdin`. This message will contain all the details necessary
+  for your implementation to start its server-under-test. Note that the request does not specify a port to listen on.
+  Implementations should instead pick an available ephemeral port according to their OS and return that value in the response.
+
+  Fields in the request are:
+
+  * `protocol` specifies what protocol will be used to run the tests. 
+  * `http_version` specifies which HTTP version will be used.
+  * `use_tls` specifies whether your server should generate a self-signed certificate. Clients will be configured to trust
+     this certificate when connecting. If `true`, the generated certificate should be returned in the `pem_cert` field of the
+     `ServerCompatResponse`. If this is set to `false`, the server should not use TLS and instead use a plaintext/unencrypted socket.
+  * `client_tls_cert` represents a PEM-encoded certificate that, if provided, clients will use to authenticate themselves. 
+     If specified, servers should require that clients provide certificates and they should ensure the presented certificate is valid.
+     Note that this will always be empty if `use_tls` is `false`.
+  * `message_receive_limit` specifies the maximum size in bytes for a message. If this value is non-zero, servers should reject
+     any message from a client that is larger than the size indicated.
+
+* Once the server is started, your implementation should then write a `ServerCompatResponse` message to `stdout`. This will
+  provide the conformance runner with details about your running server.
+
+  * `host` should be set with the host where your server is running. 
```

Comment:
This should usually be "127.0.0.1", unless the program actually starts a remote server to which the client should connect.